### PR TITLE
Add LLVM libunwind 11.0.0

### PR DESCRIPTION
**Pull request by omus**
Implements the JLL portion of: https://github.com/JuliaLang/julia/issues/30154.

TODO:
- [x] Support threading on 64-bit Windows
- [x] Verify experimental platforms

### CODE REVIEW

**Comment by omus** [comment]
I was unsuccessful in using `--deploy=local` using experimental platforms. I'll enable the settings and run it against the CI here once I verify the standard platforms work here.

**Comment by omus** [comment]
At the moment I'm working on attempting to use LLVM libunwind as a replacement for libosxunwind locally. If someone else wants to look into the Windows threading issue please go ahead.

File: L/LLVMLibUnwind/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+products = [
+    LibraryProduct("libunwind", :libunwind)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"6.1.0")
```
Comment (giordano) [inline L/LLVMLibUnwind/build_tarballs.jl:55]: This needs to require Julia v1.6:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"6.1.0", julia_compat="1.6")
```

File: L/LLVMLibUnwind/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+fi
+
+cmake ${CMAKE_FLAGS[@]}
+make -j${nprocs}
+make install
+
+# Move over the DLL. TODO: There may be a CMAKE flag for this.
+if [[ ${target} == *mingw32* ]]; then
+    mkdir -p $prefix/bin
+    mv -v lib/libunwind.dll $prefix/bin/
```
Comment (giordano) [inline L/LLVMLibUnwind/build_tarballs.jl:36]: Better use the specific environment variable for consistency:
```suggestion
    mkdir -p "${libdir}"
    mv -v lib/libunwind.dll "${libdir}"
```

File: L/LLVMLibUnwind/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+products = [
+    LibraryProduct("libunwind", :libunwind)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version = v"6.1.0", julia_compat="1.6")
```
Comment (omus) [inline L/LLVMLibUnwind/build_tarballs.jl:55]: Just a note. This is the minimum version of GCC listed by the BinaryBuilder wizard which worked for all architectures. Attempting to build with GCC 4.8.5 gave a warning that GCC 5.1 was the minimum version supported.

**Comment by giordano** [comment]
For the record, the error on Windows is
```
[  5%] Building CXX object src/CMakeFiles/unwind_static.dir/libunwind.cpp.obj
cd /workspace/srcdir/libunwind-11.0.0.src/src && /opt/bin/x86_64-w64-mingw32-libgfortran3-cxx11/x86_64-w64-mingw32-g++ --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/  -D_FILE_OFFSET_BITS=64 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS @CMakeFiles/unwind_static.dir/includes_CXX.rsp -Wa,-mbig-obj -Werror=date-time -std=gnu++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wdelete-non-virtual-dtor -Wno-comment -O3 -DNDEBUG   -Werror=return-type -W -Wall -Wchar-subscripts -Wconversion -Wmissing-braces -Wno-unused-function -Wshadow -Wsign-compare -Wsign-conversion -Wstrict-aliasing=2 -Wstrict-overflow=4 -Wunused-parameter -Wunused-variable -Wwrite-strings -Wundef -Wno-error -funwind-tables -nostdinc++ -D_DEBUG -UNDEBUG -D_LIBUNWIND_IS_NATIVE_ONLY -fno-rtti -std=c++11  -fstrict-aliasing -fno-exceptions -fno-rtti -o CMakeFiles/unwind_static.dir/libunwind.cpp.obj -c /workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp
In file included from /workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:21:0:
/workspace/srcdir/libunwind-11.0.0.src/src/AddressSpace.hpp: In static member function ‘static int64_t libunwind::LocalAddressSpace::getSLEB128(libunwind::LocalAddressSpace::pint_t&, libunwind::LocalAddressSpace::pint_t)’:
/workspace/srcdir/libunwind-11.0.0.src/src/AddressSpace.hpp:298:26: warning: conversion to ‘long long unsigned int’ from ‘int64_t {aka long long int}’ may change the sign of the result [-Wsign-conversion]
     result |= (-1ULL) << bit;
                          ^~~
/workspace/srcdir/libunwind-11.0.0.src/src/AddressSpace.hpp:298:26: warning: conversion to ‘int64_t {aka long long int}’ from ‘long long unsigned int’ may change the sign of the result [-Wsign-conversion]
In file included from /workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:73:0,
                 from /workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:22:
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp: In member function ‘bool libunwind::RWMutex::lock_shared()’:
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp:42:32: error: ‘AcquireSRWLockShared’ was not declared in this scope
     AcquireSRWLockShared(&_lock);
                                ^
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp: In member function ‘bool libunwind::RWMutex::unlock_shared()’:
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp:46:32: error: ‘ReleaseSRWLockShared’ was not declared in this scope
     ReleaseSRWLockShared(&_lock);
                                ^
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp: In member function ‘bool libunwind::RWMutex::lock()’:
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp:50:35: error: ‘AcquireSRWLockExclusive’ was not declared in this scope
     AcquireSRWLockExclusive(&_lock);
                                   ^
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp: In member function ‘bool libunwind::RWMutex::unlock()’:
/workspace/srcdir/libunwind-11.0.0.src/src/RWMutex.hpp:54:35: error: ‘ReleaseSRWLockExclusive’ was not declared in this scope
     ReleaseSRWLockExclusive(&_lock);
                                   ^
/workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp: At global scope:
/workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:30:43: warning: type-punning to incomplete type might break strict-aliasing rules [-Wstrict-aliasing]
     (unw_addr_space_t)&LocalAddressSpace::sThisAddressSpace;
                                           ^~~~~~~~~~~~~~~~~
In file included from /workspace/srcdir/libunwind-11.0.0.src/src/libunwind_ext.h:15:0,
                 from /workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:14:
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp: In instantiation of ‘bool libunwind::UnwindCursor<A, R>::getInfoFromSEH(libunwind::UnwindCursor<A, R>::pint_t) [with A = libunwind::LocalAddressSpace; R = libunwind::Registers_x86_64; libunwind::UnwindCursor<A, R>::pint_t = long long unsigned int]’:
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1905:5:   required from ‘void libunwind::UnwindCursor<A, R>::setInfoBasedOnIPRegister(bool) [with A = libunwind::LocalAddressSpace; R = libunwind::Registers_x86_64]’
/workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:323:1:   required from here
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:139:10: warning: unknown conversion type character ‘l’ in format [-Wformat=]
   fprintf(stderr, "libunwind: " msg "\n", __VA_ARGS__)
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:171:43: note: in expansion of macro ‘_LIBUNWIND_LOG’
   #define _LIBUNWIND_DEBUG_LOG(msg, ...)  _LIBUNWIND_LOG(msg, __VA_ARGS__)
                                           ^~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1811:5: note: in expansion of macro ‘_LIBUNWIND_DEBUG_LOG’
     _LIBUNWIND_DEBUG_LOG("\tpc not in table, pc=0x%llX", (uint64_t) pc);
     ^
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:139:10: warning: too many arguments for format [-Wformat-extra-args]
   fprintf(stderr, "libunwind: " msg "\n", __VA_ARGS__)
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:171:43: note: in expansion of macro ‘_LIBUNWIND_LOG’
   #define _LIBUNWIND_DEBUG_LOG(msg, ...)  _LIBUNWIND_LOG(msg, __VA_ARGS__)
                                           ^~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1811:5: note: in expansion of macro ‘_LIBUNWIND_DEBUG_LOG’
     _LIBUNWIND_DEBUG_LOG("\tpc not in table, pc=0x%llX", (uint64_t) pc);
     ^
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:139:10: warning: unknown conversion type character ‘l’ in format [-Wformat=]
   fprintf(stderr, "libunwind: " msg "\n", __VA_ARGS__)
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:171:43: note: in expansion of macro ‘_LIBUNWIND_LOG’
   #define _LIBUNWIND_DEBUG_LOG(msg, ...)  _LIBUNWIND_LOG(msg, __VA_ARGS__)
                                           ^~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1811:5: note: in expansion of macro ‘_LIBUNWIND_DEBUG_LOG’
     _LIBUNWIND_DEBUG_LOG("\tpc not in table, pc=0x%llX", (uint64_t) pc);
     ^
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:139:10: warning: too many arguments for format [-Wformat-extra-args]
   fprintf(stderr, "libunwind: " msg "\n", __VA_ARGS__)
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/config.h:171:43: note: in expansion of macro ‘_LIBUNWIND_LOG’
   #define _LIBUNWIND_DEBUG_LOG(msg, ...)  _LIBUNWIND_LOG(msg, __VA_ARGS__)
                                           ^~~~~~~~~~~~~~
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1811:5: note: in expansion of macro ‘_LIBUNWIND_DEBUG_LOG’
     _LIBUNWIND_DEBUG_LOG("\tpc not in table, pc=0x%llX", (uint64_t) pc);
     ^
In file included from /workspace/srcdir/libunwind-11.0.0.src/src/libunwind.cpp:22:0:
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1831:53: warning: conversion to ‘uint32_t {aka unsigned int}’ from ‘int’ may change the sign of the result [-Wsign-conversion]
       uint32_t lastcode = (xdata->CountOfCodes + 1) & ~1;
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/workspace/srcdir/libunwind-11.0.0.src/src/UnwindCursor.hpp:1832:33: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
       const uint32_t *handler = reinterpret_cast<uint32_t *>(&xdata->UnwindCodes[lastcode]);
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/CMakeFiles/unwind_static.dir/build.make:84: src/CMakeFiles/unwind_static.dir/libunwind.cpp.obj] Error 1
make[2]: Leaving directory '/workspace/srcdir/libunwind-11.0.0.src'
make[1]: *** [CMakeFiles/Makefile2:140: src/CMakeFiles/unwind_static.dir/all] Error 2
make[1]: Leaving directory '/workspace/srcdir/libunwind-11.0.0.src'
make: *** [Makefile:150: all] Error 2
```

**Comment by omus** [comment]
> the trick is to declare compatibility with vista for example by passing the flag `-D_WIN32_WINNT=0x0600` to the C++ compiler

Nice work!

**Comment by giordano** [comment]
I've reorganised the script to support multiple versions, in light of the possibility of having to maintain many of them at the same time

**Pull request merged**